sg/SG_Light.3
.\"
.\" Copyright (c) 2006-2022 Julien Nadeau Carriere <vedge@csoft.net>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd December 21, 2022
.Dt SG_LIGHT 3
.Os Agar 1.7
.Sh NAME
.Nm SG_Light
.Nd Agar-SG light source node
.Sh SYNOPSIS
.Bd -literal
#include <agar/core.h>
#include <agar/sg.h>
.Ed
.Sh DESCRIPTION
The
.Nm
node defines a light source in a
.Xr SG 3
scene-graph.
Each light source is associated with a priority number, which determines
which lights are used in the rendering.
The number of lights supported for rendering is dependent on the renderer
implementation.
.Sh INHERITANCE HIERARCHY
.Xr AG_Object 3 ->
.Xr SG_Node 3 ->
.Nm .
.Sh INITIALIZATION
.nr nS 1
.Ft "SG_Light *"
.Fn SG_LightNew "SG_Node *parent" "const char *name"
.Pp
.Ft void
.Fn SG_LightAmbient "SG_Light *lt" "M_Color c"
.Pp
.Ft void
.Fn SG_LightDiffuse "SG_Light *lt" "M_Color c"
.Pp
.Ft void
.Fn SG_LightSpecular "SG_Light *lt" "M_Color c"
.Pp
.Ft void
.Fn SG_LightSpot "SG_Light *lt" "M_Real cutoff" "M_Real exponent" "M_Vector3 dir"
.Pp
.nr nS 0
The
.Fn SG_LightNew
function allocates, initializes, and attaches a
.Nm
object.
.Pp
The functions
.Fn SG_LightAmbient ,
.Fn SG_LightDiffuse
and
.Fn SG_LightSpecular
configure the colors associated with the ambient, diffuse and specular
components of the light source.
.Pp
.Fn SG_LightSpot
sets the spot parameters of the light source.
.Fa cutoff
specifies the cutoff angle in degrees,
.Fa exponent
specifies the light intensity (the higher the exponent, the more focused
the light source; a value of 0 provides uniform distribution), and
.Fa dir
is the direction vector.
If the cutoff angle is < 180 degrees, the light source is directional,
otherwise it is treated as a point source and the direction vector
is ignored.
.Sh INTERNAL WIDGET API
.nr nS 1
.Ft void
.Fn SG_LightSetup "SG_Light *lt" "SG_View *sv"
.Pp
.nr nS 0
The
.Fn SG_LightSetup
function sets up the renderer states associated with a light, prior
to rendering.
This function is usually invoked from the
.Fn draw
operation of an
.Xr SG_View 3
(or derived) widget, prior to rendering a scene.
.Sh STRUCTURE DATA
For the
.Nm
object:
.Bl -tag -width "M_Real spot_exponent "
.It Ft int pri
Priority of light source when number of sources is limited (0 = highest
priority).
.It Ft M_Color ambient
Ambient color component of source.
.It Ft M_Color diffuse
Diffuse color component of source.
.It Ft M_Color specular
Specular color component of source.
.It Ft M_Real spot_cutoff
Spotlight cutoff angle (or 180 for point source).
.It Ft M_Real spot_exponent
Spotlight intensity distribution.
.It Ft M_Vector3 spot_dir
Spotlight direction.
.It Ft M_Real Kc
Constant light attenuation factor.
.It Ft M_Real Kl
Linear light attenuation factor.
.It Ft M_Real Kq
Quadratic light attenuation factor.
.El
.Sh SEE ALSO
.Xr M_Real 3 ,
.Xr M_Vector 3 ,
.Xr SG 3 ,
.Xr SG_Intro 3 ,
.Xr SG_Node 3 ,
.Xr SG_View 3
.Sh HISTORY
The
.Nm
node class first appeared in Agar 1.6.0.


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- The Agar Interface Builder [**agarib**](https://libagar.org/man1/agarib).
- The Agar Debugger [**agardb**](https://libagar.org/man1/agardb).
- Ada: Add Widget classes to the new Ada bindings in `ada/gui/`.
- New COBOL bindings: [GnuCOBOL](https://sourceforge.net/projects/open-cobol/) and [GuiCOBOL](http://www.opencobol.altervista.org).
- [**AG_Console**](https://libagar.org/man3/AG_Console): Handle ansi cursor and terminal operations. Clipboard integration.
- [**AG_FontSelector**](https://libagar.org/man3/AG_FontSelector): Display only those styles that are available for a given font (as opposed to a static list of styles).
- [**AG_Notebook**](https://libagar.org/man3/AG_Notebook): Fix padding issues. Add disposition modes Bottom, Left & Right.
- [**AG_Pixmap**](https://libagar.org/man3/AG_Pixmap) & [**AG_Fixed**](https://libagar.org/man3/AG_Fixed): Zoom operations.
- [**AG_StyleEditor**](https://libagar.org/man3/AG_StyleEditor): Edition of stylesheet rules.
- [**AG_StyleSheet**](https://libagar.org/man3/AG_StyleSheet): Gradients and border styles. New selector "Class > #id" for specific named instances of attached child widgets.
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Filters. Stencil operations.
- [**AG_Text**](https://libagar.org/man3/AG_Text): Provide a way to map core font names to known system fonts so porters have the option of adding dependencies to existing font packages in order to reduce the size of agar packages.
- [**AG_Textbox**](https://libagar.org/man3/AG_Textbox) & [**AG_Editable**](https://libagar.org/man3/AG_Editable): Extend SGR support. Syntax highlighting & rich-text editing methods.
- [**AG_WidgetPrimitives**](https://libagar.org/man3/AG_WidgetPrimitives): Dithering. Shadow effects.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Provide a variation of the "zoom" feature to resize widgets individually.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Threading optimizations. Cache the `VISIBLE` flag into an `int` so that culling / rendering and mouse-intersection tests can skip over invisible widgets without locking them.
- [**AG_Window**](https://libagar.org/man3/AG_Window): MRU API to simplify the process of remembering geometries. New gravity methods for autoplacing.
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Cursor operations. Clipboard integration.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Make drag-and-drop (`sockets` test) work correctly (possibly using `AG_WindowMove()`).
- [**wgl**](https://libagar.org/man3/AG_DriverWGL): Clipboard integration. Live resize.
- Fix [**Style Editor**](https://libagar.org/man3/AG_StyleEditor) and [**GUI Debugger**](https://libagar.org/man3/AG_GuiDebugger) crashes which would occur when attempting to iterate over widgets (in keyrepeat mode).
- Build system: Add [**CMake**](https://cmake.org) support.
- A software framebuffer driver for X Windows.

## [1.7.0] - 2022-10-
### Added
- [**sdl2fb**](https://libagar.org/man3/AG_DriverSDL2FB): New driver for SDL 2.0 (single-window; frame-buffer mode).
- [**sdl2gl**](https://libagar.org/man3/AG_DriverSDL2GL): New driver for SDL 2.0 (single-window; OpenGL mode).
- [**sdl2mw**](https://libagar.org/man3/AG_DriverSDL2MW): New driver for SDL 2.0 (multi-window; OpenGL mode).
- [**Single-window drivers**](https://libagar.org/man3/AG_DriverSw): New option `clampOnResize` (clamp active Agar windows against new display size).
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Implemented keyboard shortcuts. Added `AG_MENU_NO_BOOL_MSG` and `AG_MENU_FAST_BOOL_MSG` options.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): Improved the performance of the `AG_MenuView` renderer by caching state-dependent opaque labels (embedding any icons), so blending is no longer required. Improved appearance of `AG_MenuView`.
- [**AG_Menu**](https://libagar.org/man3/AG_Menu): When an `AG_MenuView` is detached, immediately invalidate the `view` pointers of its `AG_MenuItem`.
- [**AG_Radio**](https://libagar.org/man3/AG_Radio): New `HOMOGENOUS` option to divide space equally between items. Thanks Stephen!
- [**AG_Surface**](https://libagar.org/man3/AG_Surface): Embed 4 general-purpose 16-bit guides. Guides are useful for typography, graphics and other applications.
- [**AG_Text**](https://libagar.org/man3/AG_Text): The surface returned by `AG_TextRender()` now includes Ascent information under `Guide 0`.
- [**AG_Textbox**](https://libagar.org/man3/AG_Textbox) and [**AG_Editable**](https://libagar.org/man3/AG_Editable): Implement Undo/Redo functionality.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Draw lines to express the relationship between tree items.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_EXPAND_NODES`. Display nodes (items with children) initially in expanded state.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_NO_SCALE_ICON`. Disable icon scaling when the icon width exceeds the standard width.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New option `AG_TLIST_FIXED_HEIGHT`. Don't reset the item height on "font-changed".
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New item flag `AG_TLIST_ITEM_DISABLED`. Disable the item individually (prevent selection and display in "disabled" state).
- [**AG_Event**](https://libagar.org/man3/AG_Event): New macro [AG_EVENT_DUMP](https://libagar.org/man3/AG_EVENT_DUMP) to produce a listing of `argv[]` arguments on the console.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): New flag `AG_WIDGET_DISABLE_ON_ATTACH` and functions `AG_PushDisabledState()` and `AG_PopDisabledState`. Make attached child widgets start in DISABLED mode (without raising "widget-disabled").

sg/sg_light.c
/*
 * Copyright (c) 2006-2019 Julien Nadeau Carriere <vedge@csoft.net>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Light source node object.
 */

#include <agar/core/core.h>
#include <agar/sg/sg.h>
#include <agar/sg/sg_gui.h>

#include <string.h>

SG_Light *
SG_LightNew(void *parent, const char *name)
{
	SG_Light *lt;

	lt = Malloc(sizeof(SG_Light));
	AG_ObjectInit(lt, &sgLightClass);
	if (name) {
		AG_ObjectSetNameS(lt, name);
	} else {
		OBJECT(lt)->flags |= AG_OBJECT_NAME_ONATTACH;
	}
	AG_ObjectAttach(parent, lt);
	return (lt);
}

static void
Init(void *_Nonnull obj)
{
	SG_Light *lt = obj;
	static int nlight = GL_LIGHT0;			/* XXX */

	lt->flags = 0;
	lt->pri = 0;
	lt->light = nlight++;				/* XXX test */
	lt->tag = 0;
	lt->ambient = M_ColorBlack();
	lt->diffuse = M_ColorWhite();
	lt->specular = M_ColorWhite();
	lt->spot_dir = M_VecZero3();
	lt->spot_exponent = 3.0;
	lt->spot_cutoff = 180.0;
	lt->Kc = 1.0;
	lt->Kl = 0.0;
	lt->Kq = 0.0;
#ifdef HAVE_GLU
	lt->qo = gluNewQuadric();
	gluQuadricDrawStyle(lt->qo, GLU_FILL);
	gluQuadricNormals(lt->qo, GLU_SMOOTH);
#endif
}

static void
Destroy(void *_Nonnull obj)
{
#ifdef HAVE_GLU
	SG_Light *lt = obj;
	gluDeleteQuadric(lt->qo);
#endif
}

static int
Load(void *_Nonnull obj, AG_DataSource *_Nonnull ds, const AG_Version *_Nonnull ver)
{
	SG_Light *lt = obj;

	lt->flags = (Uint)AG_ReadUint8(ds);
	lt->pri = (int)AG_ReadSint32(ds);
	lt->light = (int)AG_ReadSint32(ds);
	lt->tag = (int)AG_ReadSint32(ds);
	lt->ambient = M_ReadColor(ds);
	lt->diffuse = M_ReadColor(ds);
	lt->specular = M_ReadColor(ds);
	lt->spot_dir = M_ReadVector3(ds);
	lt->spot_exponent = M_ReadReal(ds);
	lt->spot_cutoff = M_ReadReal(ds);
	lt->Kc = M_ReadReal(ds);
	lt->Kl = M_ReadReal(ds);
	lt->Kq = M_ReadReal(ds);
	return (0);
}

static int
Save(void *_Nonnull obj, AG_DataSource *_Nonnull ds)
{
	SG_Light *lt = obj;

	AG_WriteSint32(ds, (Sint32)lt->pri);
	AG_WriteSint32(ds, (Sint32)lt->light);
	AG_WriteSint32(ds, (Sint32)lt->tag);
	M_WriteColor(ds, &lt->ambient);
	M_WriteColor(ds, &lt->diffuse);
	M_WriteColor(ds, &lt->specular);
	M_WriteVector3(ds, &lt->spot_dir);
	M_WriteReal(ds, lt->spot_exponent);
	M_WriteReal(ds, lt->spot_cutoff);
	M_WriteReal(ds, lt->Kc);
	M_WriteReal(ds, lt->Kl);
	M_WriteReal(ds, lt->Kq);
	return (0);
}

static void
Draw(void *_Nonnull obj, SG_View *_Nonnull view)
{
#ifdef HAVE_GLU
	SG_Light *lt = obj;

	if (lt->spot_cutoff == 180.0) {
		gluSphere(lt->qo, 0.125, 4, 4);
	} else {
		gluCylinder(lt->qo, 0.125, 0.125, 1.0, 4, 3); 
	}
#endif /* HAVE_GLU */
}

/*
 * Set up OpenGL light sources for rendering.
 * Must be called from widget draw context.
 */
void
SG_LightSetup(SG_Light *lt)
{
	const int which = lt->light;
	M_Vector3 v;
	GLfloat posf[4];
	GLfloat dirf[4];

	AG_ObjectLock(lt);

	GL_Enable(which);
	v = SG_NodePos(lt);
	posf[0] = (GLfloat)v.x;
	posf[1] = (GLfloat)v.y;
	posf[2] = (GLfloat)v.z;
	posf[3] = 1.0f;
	glLightfv(which, GL_POSITION, posf);

#ifdef DOUBLE_PRECISION
	{
		float ambient[4];
		float diffuse[4];
		float specular[4];

		M_ColorTo4fv(&lt->ambient, ambient);
		glLightfv(which, GL_AMBIENT, ambient);
		M_ColorTo4fv(&lt->diffuse, diffuse);
		glLightfv(which, GL_DIFFUSE, diffuse);
		M_ColorTo4fv(&lt->specular, specular);
		glLightfv(which, GL_SPECULAR, specular);

		if (lt->spot_cutoff != 180.0) {
			v = SG_NodeDir(lt);
			dirf[0] = (GLfloat)v.x;
			dirf[1] = (GLfloat)v.y;
			dirf[2] = (GLfloat)v.z;
			dirf[3] = 1.0f;
			glLightfv(which, GL_SPOT_DIRECTION, dirf);
		}
	}
#else /* !DOUBLE_PRECISION */
	glLightfv(which, GL_AMBIENT,  (GLfloat *)&lt->ambient);
	glLightfv(which, GL_DIFFUSE,  (GLfloat *)&lt->diffuse);
	glLightfv(which, GL_SPECULAR, (GLfloat *)&lt->specular);
	if (lt->spot_cutoff != 180.0) {
		glLightfv(which, GL_SPOT_DIRECTION, dirf);
	}
#endif
	glLightf(which, GL_SPOT_EXPONENT,         (GLfloat)lt->spot_exponent);
	glLightf(which, GL_SPOT_CUTOFF,           (GLfloat)lt->spot_cutoff);
	glLightf(which, GL_CONSTANT_ATTENUATION,  (GLfloat)lt->Kc);
	glLightf(which, GL_LINEAR_ATTENUATION,    (GLfloat)lt->Kl);
	glLightf(which, GL_QUADRATIC_ATTENUATION, (GLfloat)lt->Kq);
	
	AG_ObjectUnlock(lt);
}

static void
SelectColor(AG_Event *_Nonnull event)
{
	AG_HSVPal *pal = AG_PTR(1);
	void *color = AG_PTR(2);
	AG_Mutex *lock = AG_PTR(3);

	M_BindRealMp(pal, "RGBAv", color, lock);
}

static void *_Nullable
Edit(void *_Nonnull obj, SG_View *_Nullable sgv)
{
	SG_Light *lt = obj;
	AG_Mutex *lock = &OBJECT(lt)->lock;
	AG_Notebook *nb;
	AG_NotebookTab *ntab;
	AG_HSVPal *pal;
	AG_Numerical *num;

	nb = AG_NotebookNew(NULL, AG_NOTEBOOK_EXPAND);
	ntab = AG_NotebookAdd(nb, _("Src"), AG_BOX_VERT);
	{
		M_EditTranslate3Mp(ntab, _("Position: "),
		    &SGNODE(lt)->T, lock);

		num = AG_NumericalNew(ntab, 0, NULL, _("Priority: "));
		AG_BindIntMp(num, "value", &lt->pri, lock);

		AG_SeparatorNewHoriz(ntab);
		
		AG_LabelNew(ntab, 0, _("Using: GL_LIGHT%i"),
		    (int)(lt->light - GL_LIGHT0));

		AG_SeparatorNewHoriz(ntab);

		num = AG_NumericalNew(ntab, 0, "deg", _("Cutoff angle: "));
		M_BindRealMp(num, "value", &lt->spot_cutoff, lock);

		num = AG_NumericalNew(ntab, 0, NULL, _("Spot exponent: "));
		M_BindRealMp(num, "value", &lt->spot_exponent, lock);
		M_SetReal(num, "inc", 0.1);
	}
	ntab = AG_NotebookAdd(nb, _("Attenuation"), AG_BOX_VERT);
	{
		num = AG_NumericalNew(ntab, 0, NULL, "Kc: ");
		M_BindRealMp(num, "value", &lt->Kc, lock);
		M_SetReal(num, "inc", 0.01);
	
		num = AG_NumericalNew(ntab, 0, NULL, "Kl: ");
		M_BindRealMp(num, "value", &lt->Kl, lock);
		M_SetReal(num, "inc", 0.001);

		num = AG_NumericalNew(ntab, 0, NULL, "Kq: ");
		M_BindRealMp(num, "value", &lt->Kq, lock);
		M_SetReal(num, "inc", 0.00001);
	}
	ntab = AG_NotebookAdd(nb, _("Color"), AG_BOX_VERT);
	{
		AG_Toolbar *bar;
		
		pal = AG_HSVPalNew(ntab, AG_HSVPAL_EXPAND);
		M_BindRealMp(pal, "RGBAv", (void *)&lt->ambient, lock);

		bar = AG_ToolbarNew(ntab, AG_TOOLBAR_HORIZ, 1,
		    AG_TOOLBAR_HOMOGENOUS|AG_TOOLBAR_STICKY);
		{
			AG_ToolbarButton(bar, _("Ambient"), 1,
			    SelectColor, "%p,%p", pal, &lt->ambient, lock);
			AG_ToolbarButton(bar, _("Diffuse"), 0,
			    SelectColor, "%p,%p", pal, &lt->diffuse, lock);
			AG_ToolbarButton(bar, _("Specular"), 0,
			    SelectColor, "%p,%p", pal, &lt->specular, lock);
		}
	}
	return (nb);
}

#if 0
static void
MenuInstance(void *_Nonnull obj, AG_MenuItem *_Nonnull m, SG_View *_Nonnull sgv)
{
#if 0
	SG_Light *lt = obj;

	AG_MenuAction(m, _("Light parameters..."), NULL,
	    EditLightParams, "%p", lt);
#endif
}
#endif

SG_NodeClass sgLightClass = {
	{
		"SG_Node:SG_Light",
		sizeof(SG_Light),
		{ 0,0 },
		Init,
		NULL,	/* reset */
		Destroy,
		Load,
		Save,
		SG_NodeEdit
	},
	NULL,			/* menuInstance */
	NULL,			/* menuClass */
	Draw,
	NULL,			/* intersect */
	Edit
};
